tests/unit/test_smtp_observer.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""SMTPObserver unit tests."""
+
+# pylint: disable=protected-access
+
+import pytest
+from charms.smtp_integrator.v0.smtp import AuthType, TransportSecurity
+from ops.testing import Harness
+
+from charm_types import SMTPConfiguration
+
+RELATION_DATA_TO_SMTP_CONF_PARAMS = [
+    (
+        {
+            "host": "127.0.0.1",
+            "port": "25",
+            "auth_type": AuthType.NONE,
+            "transport_security": TransportSecurity.NONE,
+        },
+        SMTPConfiguration(
+            enable_tls=False,
+            force_tls=False,
+            require_transport_security=False,
+            host="127.0.0.1",
+            port=25,
+            user=None,
+            password=None,
+        ),
+    ),
+    (
+        {
+            "host": "127.0.0.1",
+            "port": "25",
+            "auth_type": AuthType.PLAIN,
+            "transport_security": TransportSecurity.STARTTLS,
+            "user": "username",
+            "password": "SECRET",  # no sec
+        },
+        SMTPConfiguration(
+            enable_tls=True,
+            force_tls=False,
+            require_transport_security=True,
+            host="127.0.0.1",
+            port=25,
+            user="username",
+            password="SECRET",  # nosec
+        ),
+    ),
+    (
+        {
+            "host": "127.0.0.1",
+            "port": "587",
+            "auth_type": AuthType.PLAIN,
+            "transport_security": TransportSecurity.TLS,
+            "user": "username",
+            "password": "SECRET",  # nosec
+        },
+        SMTPConfiguration(
+            enable_tls=True,
+            force_tls=True,
+            require_transport_security=True,
+            host="127.0.0.1",
+            port=587,
+            user="username",
+            password="SECRET",  # nosec
+        ),
+    ),
+]
+
+
+@pytest.mark.parametrize("relation_data, expected_config", RELATION_DATA_TO_SMTP_CONF_PARAMS)
```

Comment:
You might be interested in taking a look at the following pattern for how to parametrize tests, do you think that would be helpful here? https://github.com/canonical/discourse-gatekeeper/blob/d49ebe76aaec9fc64947a13728edbb650e69048d/tests/unit/test_index.py#L183

Reply:
Thanks for the info. Changed to a function with `pytest.param`.